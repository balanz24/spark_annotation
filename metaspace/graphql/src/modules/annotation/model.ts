import {
  Entity,
  PrimaryColumn,
  Column, ManyToOne, JoinColumn, OneToMany, Index, PrimaryGeneratedColumn,
} from 'typeorm'
import { MomentValueTransformer } from '../../utils/MomentValueTransformer'
import { Moment } from 'moment'
import { MolecularDB } from '../moldb/model'

@Entity('coloc_job')
export class ColocJob {
  @PrimaryColumn({ type: 'uuid', default: () => 'uuid_generate_v1mc()' })
  id: string;

  @Column({ type: 'text', name: 'ds_id' })
  datasetId: string;

  @Column({ type: 'int' })
  moldbId: number;

  @Column({ type: 'numeric', precision: 2, scale: 2 })
  fdr: number;

  @Column({ type: 'text' })
  algorithm: string;

  @Column({ type: 'timestamp without time zone', transformer: new MomentValueTransformer() })
  start: Moment;

  @Column({ type: 'timestamp without time zone', transformer: new MomentValueTransformer() })
  finish: Moment;

  @Column({ type: 'text', nullable: true })
  error: string | null;

  @Column({ type: 'int', array: true })
  sampleIonIds: number[];

  @OneToMany(() => ColocAnnotation, colocAnnotation => colocAnnotation.colocJob)
  colocAnnotations: ColocAnnotation[];

  @ManyToOne(() => MolecularDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'moldb_id' })
  molecularDB: MolecularDB;
}

@Entity('coloc_annotation')
export class ColocAnnotation {
  @PrimaryColumn({ type: 'uuid' })
  colocJobId: string;

  /** Index of the annotation's sfAdduct in the parent ColocJob's orderedMols array */
  @PrimaryColumn({ type: 'int' })
  ionId: number;

  @Column({ type: 'int', array: true })
  colocIonIds: number[];

  @Column({ type: 'float4', array: true })
  colocCoeffs: number[];

  @ManyToOne(() => ColocJob, { onDelete: 'CASCADE' })
  @JoinColumn()
  colocJob: ColocJob;
}

@Entity('ion')
@Index(['formula', 'chemMod', 'neutralLoss', 'adduct', 'charge'], { unique: true })
export class Ion {
  @PrimaryGeneratedColumn({ type: 'int' })
  id: number;

  /**
   * All components of the ion in one string, with charge as a '+'/'-' suffix e.g. "C21H41O6P-HO+HO3P-H2O-CO2+Na+".
   * Note that this isn't unique - it's possible for different combinations of chemMods, neutralLosses and adducts
   * to build the same ion.
   */
  @Column({ type: 'text' })
  @Index()
  ion: string;

  /** The formula of the molecule prior to any modifications, e.g. "C21H41O6P".
   *  Internally this has been called "sumFormula" or "sf", but this term is incorrect.
   *  It should be referred to as "formula" or "molecular formula". */
  @Column({ type: 'text' })
  formula: string;

  @Column({ type: 'text', default: '' })
  chemMod: string;

  @Column({ type: 'text', default: '' })
  neutralLoss: string;

  @Column({ type: 'text' })
  adduct: string;

  @Column({ type: 'text', default: '' })
  ionFormula: string;

  /** 1 or -1 */
  @Column({ type: 'smallint' })
  charge: number;
}

export const ANNOTATION_ENTITIES = [
  ColocJob,
  ColocAnnotation,
  Ion,
]
