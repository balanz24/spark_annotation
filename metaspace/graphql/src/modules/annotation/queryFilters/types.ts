import {
  AnnotationFilter,
  AnnotationOrderBy,
  DatasetFilter,
  DatasetOrderBy,
  SortingOrder,
} from '../../../binding'
import { ESAnnotation } from '../../../../esConnector'

// A superset of allAnnotations, countAnnotations, allDatasets and countDatasets
export interface QueryFilterArgs {
  orderBy?: AnnotationOrderBy | DatasetOrderBy;
  sortingOrder?: SortingOrder;
  offset?: number;
  limit?: number;
  filter?: AnnotationFilter;
  datasetFilter?: DatasetFilter;
  simpleQuery?: string;
}

export type PostProcessFunc = (annotations: ESAnnotation[]) => ESAnnotation[];

export interface QueryFilterResult {
  args: QueryFilterArgs;
  postprocess?: PostProcessFunc;
}

export interface ESAnnotationWithColoc extends ESAnnotation {
  _cachedColocCoeff: number | null;
  _isColocReference: boolean;

  getColocalizationCoeff(_colocalizedWith: string, _colocalizationAlgo: string,
                         _databaseId: number, _fdrLevel: number | null): number | null;
}
