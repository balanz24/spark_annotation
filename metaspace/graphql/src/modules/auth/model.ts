import 'reflect-metadata'
import { Column, Entity, PrimaryColumn } from 'typeorm'
import { Moment } from 'moment'
import { MomentValueTransformer } from '../../utils/MomentValueTransformer'

@Entity()
export class Credentials {
  @PrimaryColumn({ type: 'uuid', default: () => 'uuid_generate_v1mc()' })
  id: string;

  @Column({ type: 'text', nullable: true })
  hash: string | null;

  @Column({ type: 'text', nullable: true })
  googleId: string | null;

  @Column({ type: 'text', nullable: true })
  apiKey: string | null;

  @Column({ type: 'timestamp without time zone', nullable: true, transformer: new MomentValueTransformer() })
  apiKeyLastUpdated: Moment | null;

  @Column({ type: 'text', nullable: true })
  emailVerificationToken: string | null;

  @Column({
    type: 'timestamp without time zone',
    nullable: true,
    transformer: new MomentValueTransformer(),
  })
  emailVerificationTokenExpires?: Moment | null;

  @Column({ type: 'boolean', default: false })
  emailVerified: boolean;

  @Column({ type: 'text', nullable: true })
  resetPasswordToken: string | null;

  @Column({
    type: 'timestamp without time zone',
    nullable: true,
    transformer: new MomentValueTransformer(),
  })
  resetPasswordTokenExpires: Moment | null;
}

export const AUTH_ENTITIES = [
  Credentials,
]
