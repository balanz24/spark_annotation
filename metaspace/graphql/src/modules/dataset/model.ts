import { Column, Entity, JoinColumn, ManyToOne, OneToMany, OneToOne, PrimaryColumn } from 'typeorm'
import { Group } from '../group/model'
import { User } from '../user/model'
import { Project } from '../project/model'
import { ExternalLink } from '../project/ExternalLink'
import { DatasetDiagnostic, EngineDataset } from '../engine/model'
import { EnrichmentBootstrap } from '../enrichmentdb/model'

@Entity()
export class Dataset {
  @PrimaryColumn({ type: 'text' })
  id: string;

  @Column({ type: 'uuid' })
  userId: string; // dataset submitter and owner -> edit rights

  @ManyToOne(() => User, user => user.datasets)
  @JoinColumn({ name: 'user_id' })
  user: User;

  @Column({ type: 'text', nullable: true })
  description: string | null;

  @Column({ type: 'uuid', nullable: true })
  groupId: string | null; // dataset belongs to group -> all members have view rights

  @ManyToOne(() => Group)
  @JoinColumn({ name: 'group_id' })
  group: Group;

  @Column({ default: false })
  groupApproved: boolean; // true when submitter is a group member

  // when user manually inputs PI details
  @Column({ type: 'text', nullable: true })
  piName: string | null;

  @Column({ type: 'text', nullable: true })
  piEmail: string | null;

  // sm-engine and sm-graphql create & manage the public.dataset (EngineDataset) and graphql.dataset (Dataset) tables
  // independently, so this relationship doesn't enforce an FK.
  @OneToOne(() => EngineDataset, { createForeignKeyConstraints: false })
  @JoinColumn({ name: 'id' })
  engineDataset: EngineDataset;

  @OneToMany(() => DatasetProject, datasetProject => datasetProject.dataset)
  datasetProjects: DatasetProject[];

  @Column({ type: 'json', nullable: true })
  externalLinks: ExternalLink[] | null;

  @OneToMany(() => DatasetDiagnostic, datasetDiagnostic => datasetDiagnostic.dataset)
  datasetDiagnostics: DatasetDiagnostic[];

  @OneToMany(() => EnrichmentBootstrap, enrichmentBootstrap => enrichmentBootstrap.dataset)
  enrichmentBootstrap: EnrichmentBootstrap[];
}

@Entity({ name: 'dataset_project' })
export class DatasetProject {
  @PrimaryColumn({ type: 'text' })
  datasetId: string;

  @ManyToOne(() => Dataset)
  @JoinColumn({ name: 'dataset_id' })
  dataset: Dataset;

  @PrimaryColumn({ type: 'uuid' })
  projectId: string;

  @ManyToOne(() => Project)
  @JoinColumn({ name: 'project_id' })
  project: Project;

  @Column({ type: 'boolean' })
  approved: boolean;
}

export const DATASET_ENTITIES = [
  Dataset,
  DatasetProject,
]
