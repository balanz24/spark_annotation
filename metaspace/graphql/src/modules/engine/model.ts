import {
  Column,
  Entity,
  Index,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryColumn,
  PrimaryGeneratedColumn,
  Unique,
} from 'typeorm'
import { MomentValueTransformer } from '../../utils/MomentValueTransformer'
import { Ion } from '../annotation/model'
import { MolecularDB } from '../moldb/model'
import { Dataset } from '../dataset/model'
import { Moment } from 'moment'
import { EnrichmentBootstrap } from '../enrichmentdb/model'

export type DatasetStatus = 'QUEUED' | 'ANNOTATING' | 'FINISHED' | 'FAILED';

@Entity({ schema: 'public', name: 'dataset' })
export class EngineDataset {
  @PrimaryColumn({ type: 'text' })
  id: string;
  @Index('ind_dataset_name')
  @Column({ type: 'text', nullable: true })
  name: string | null;
  @Column({ type: 'text', nullable: true })
  inputPath: string | null;
  @Column({ type: 'json', nullable: true })
  metadata: any;
  @Column({ type: 'json', nullable: true })
  roi: any;
  @Column({ type: 'json', nullable: true })
  config: any;
  @Column({ type: 'timestamp without time zone', nullable: true, transformer: new MomentValueTransformer() })
  uploadDt: Date | null;
  @Column({ type: 'text', nullable: true })
  status: DatasetStatus | null;
  @Column({ type: 'timestamp without time zone', transformer: new MomentValueTransformer() })
  statusUpdateDt: Date | null;
  @Column({ type: 'text', nullable: true })
  opticalImage: string | null;
  @Column({ type: 'double precision', array: true, nullable: true })
  transform: number[] | null;
  @Column({ type: 'boolean', default: true })
  isPublic: boolean;
  @Column({ type: 'json', nullable: true })
  acqGeometry: any;
  @Column({ type: 'text', default: 'fs' })
  ionImgStorageType: string;
  @Column({ type: 'text', nullable: true })
  thumbnail: string | null;
  @Column({ type: 'text', nullable: true })
  thumbnailUrl: string | null;
  @Column({ type: 'text', nullable: true })
  ionThumbnail: string | null;
  @Column({ type: 'text', nullable: true })
  ionThumbnailUrl: string | null;
  @Column({ type: 'json', nullable: true })
  sizeHash: any;

  // sm-engine and sm-graphql create & manage the public.dataset (EngineDataset) and graphql.dataset (Dataset) tables
  // independently, so this relationship doesn't enforce an FK.
  @ManyToOne(() => Dataset, { createForeignKeyConstraints: false })
  @JoinColumn({ name: 'id' })
  dataset: Dataset;

  @OneToMany(() => OpticalImage, opticalImage => opticalImage.dataset)
  opticalImages: OpticalImage[];

  @OneToMany(() => Job, job => job.dataset)
  jobs: Job[];

  @OneToMany(() => PerfProfile, pipelineStats => pipelineStats.dataset)
  pipelineStats: PerfProfile[];

  @OneToMany(() => DatasetDiagnostic, datasetDiagnostic => datasetDiagnostic.engineDataset)
  datasetDiagnostics: DatasetDiagnostic[];

  @OneToMany(() => EnrichmentBootstrap, enrichmentBootstrap => enrichmentBootstrap.engineDataset)
  enrichmentBootstrap: EnrichmentBootstrap[];
}

@Entity({ schema: 'public' })
export class OpticalImage {
  @PrimaryColumn({ type: 'text' })
  id: string;
  @Column({ name: 'ds_id' })
  datasetId: string;
  @Column({ type: 'text' })
  type: string;
  @Column({ type: 'real' })
  zoom: number;
  @Column({ type: 'int' })
  width: number;
  @Column({ type: 'int' })
  height: number;
  @Column({ type: 'real', array: true, nullable: true })
  transform: number[][];
  @Column({ type: 'text', nullable: true })
  url: string;

  @ManyToOne(() => EngineDataset, dataset => dataset.opticalImages, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'ds_id' })
  dataset: EngineDataset;
}

@Entity({ schema: 'public' })
export class Job {
  @PrimaryGeneratedColumn()
  id: number;
  @Column({ name: 'moldb_id', type: 'int' })
  moldbId: number;
  @Column({ name: 'ds_id', nullable: true })
  datasetId: string | null;
  @Column({ type: 'text', nullable: true })
  status: DatasetStatus | null;
  @Column({ type: 'timestamp', nullable: true })
  start: Date | null;
  @Column({ type: 'timestamp', nullable: true })
  finish: Date | null;

  @ManyToOne(() => EngineDataset, dataset => dataset.jobs, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'ds_id' })
  dataset: EngineDataset;

  @ManyToOne(() => MolecularDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'moldb_id' })
  molecularDB: MolecularDB;

  @OneToMany(() => Annotation, annotation => annotation.job)
  annotations: Annotation[];

  @OneToMany(() => DatasetDiagnostic, datasetDiagnostic => datasetDiagnostic.job)
  datasetDiagnostics: DatasetDiagnostic[];
}

// Should match the literal in metaspace/engine/sm/engine/annotation/diagnostics.py
export type DiagnosticType = 'TIC' | 'IMZML_METADATA' | 'FDR_RESULTS'
export const DiagnosticTypeOptions: {[k in DiagnosticType]: k} = {
  TIC: 'TIC',
  IMZML_METADATA: 'IMZML_METADATA',
  FDR_RESULTS: 'FDR_RESULTS',
}

export type DiagnosticImageFormat = 'PNG' | 'NPY' | 'JSON' | 'PARQUET'
export interface DiagnosticImage {
  key: string;
  index?: number;
  image_id: string;
  url: string;
  format: DiagnosticImageFormat;
}

@Entity({ schema: 'public' })
// This is the main index for lookup performance. It enforces uniqueness when jobId is non-null, however because
// NULL != NULL in postgres, this index doesn't enforce uniqueness of (datasetId, type) when jobId is NULL.
// (i.e. counterintuitively, this index won't prevent inserting ('2001-01-01_00h00m00s', 'TIC', NULL) even if an
// identical row already exists, because NULL != NULL, so NULL is always considered a unique value)
// Therefore a second index is used for enforcing uniqueness when jobId is null.
@Index(['datasetId', 'type', 'jobId'], { unique: true })
@Index(['datasetId', 'type'], { unique: true, where: 'job_id IS NULL' })
export class DatasetDiagnostic {
  @PrimaryColumn({ type: 'uuid', default: () => 'uuid_generate_v1mc()' })
  id: string;

  @Column({ type: 'text', enum: Object.values(DiagnosticTypeOptions) })
  type: DiagnosticType;

  @Column({ name: 'ds_id', type: 'text' })
  datasetId: string;

  @Column({ type: 'int', nullable: true })
  jobId: number | null;

  @Column({
    type: 'timestamp without time zone',
    default: () => "(now() at time zone 'utc')",
    transformer: new MomentValueTransformer(),
  })
  updatedDT: Moment;

  @Column({ type: 'json', nullable: true })
  data: any;

  @Column({ type: 'text', nullable: true })
  error: any;

  // images are stored separately in an array with a defined structure so that it's easy for all images to be cleaned
  // up when a dataset is deleted, without needing to understand any of the structure in `data`.
  // All created Image IDs MUST be included in this column. Additionally, they MAY be referenced in `data`
  // (preferably by `key` instead of `image_id`).
  @Column({ type: 'json' })
  images: DiagnosticImage[];

  // This table is a child table of public.dataset, not graphql.dataset, so avoid making an FK
  @ManyToOne(() => Dataset, { createForeignKeyConstraints: false })
  @JoinColumn({ name: 'ds_id' })
  dataset: Dataset;

  @ManyToOne(() => EngineDataset, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'ds_id' })
  engineDataset: EngineDataset;

  @ManyToOne(() => Job, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'job_id' })
  job: Job;
}

interface AnnotationStats {
  chaos: number;
  spatial: number;
  spectral: number;
  msm: number;
  total_iso_ints: number[];
  min_iso_ints: number[];
  max_iso_ints: number[];
}

type OffSampleLabel = 'off' | 'on';
interface AnnotationOffSample {
  prob: number;
  label: OffSampleLabel;
}

@Entity({ schema: 'public' })
@Unique('annotation_annotation_uindex', ['jobId', 'formula', 'chemMod', 'neutralLoss', 'adduct'])
export class Annotation {
  @PrimaryGeneratedColumn()
  id: number;
  @Index('annotation_job_id_index')
  @Column()
  jobId: number;
  @Column({ type: 'text' })
  formula: string;
  @Column({ type: 'text' })
  chemMod: string;
  @Column({ type: 'text' })
  neutralLoss: string;
  @Column({ type: 'text' })
  adduct: string;
  @Column({ type: 'real' })
  msm: number;
  @Column({ type: 'real' })
  fdr: number;
  @Column({ type: 'json' })
  stats: AnnotationStats;
  @Column({ type: 'text', array: true })
  isoImageIds: string[];
  @Column({ type: 'json', nullable: true })
  offSample: AnnotationOffSample | null;
  @Column({ type: 'int', nullable: true })
  ionId: number | null;

  @ManyToOne(() => Job, job => job.annotations, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'job_id' })
  job: Job;

  @ManyToOne(() => Ion, { onDelete: 'SET NULL' })
  @JoinColumn({ name: 'ion_id' })
  ion: Ion | null;
}

@Entity({ schema: 'public' })
export class PerfProfile {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: 'text' })
  taskType: string;

  @Column({ name: 'ds_id', nullable: true })
  datasetId: string | null;

  @Column({ type: 'timestamp' })
  start: Date;

  @Column({ type: 'timestamp', nullable: true })
  finish: Date | null;

  @Column({ type: 'json', nullable: true })
  extraData: any;

  @Column({ type: 'text', nullable: true })
  logs: string | null;

  @Column({ type: 'text', nullable: true })
  error: string | null;

  @ManyToOne(() => EngineDataset, dataset => dataset.pipelineStats, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'ds_id' })
  dataset: EngineDataset;

  @OneToMany(() => PerfProfileEntry, entry => entry.profile)
  entries: PerfProfileEntry[];
}

@Entity({ schema: 'public' })
export class PerfProfileEntry {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: 'int' })
  profileId: number;

  @Column({ type: 'int' })
  sequence: number;

  @Column({ type: 'timestamp' })
  start: Date;

  @Column({ type: 'timestamp' })
  finish: Date;

  @Column({ type: 'text' })
  name: string;

  @Column({ type: 'json', nullable: true })
  extraData: any;

  @ManyToOne(() => PerfProfile, profile => profile.entries, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'profile_id' })
  profile: PerfProfile;
}

@Entity({ schema: 'public' })
export class ScoringModel {
  @PrimaryGeneratedColumn()
  id: number;

  /** `name` is for finding models based on a dataset's `scoring_model` DSConfig value */
  @Index({ unique: true })
  @Column({ type: 'text' })
  name: string;

  /** `type` is hint for how to interpret `params`. sm-engine is responsible for managing the valid values. */
  @Column({ type: 'text' })
  type: string;

  /** `params` is a JSON blob of anything sm-engine needs to load & use the model. It's intentionally schemaless,
   as it's very likely that future model types will need to extend the structure. */
  @Column({ type: 'json' })
  params: any;
}

export const ENGINE_ENTITIES = [
  EngineDataset,
  OpticalImage,
  Job,
  DatasetDiagnostic,
  Annotation,
  PerfProfile,
  PerfProfileEntry,
  ScoringModel,
]
