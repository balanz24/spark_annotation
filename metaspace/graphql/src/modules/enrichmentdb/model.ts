import {
  Column,
  Entity,
  Index,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
} from 'typeorm'
import { MolecularDB } from '../moldb/model'
import { Dataset } from '../dataset/model'
import { EngineDataset } from '../engine/model'
import { MomentValueTransformer } from '../../utils/MomentValueTransformer'

@Entity({ schema: 'public' })
export class EnrichmentDB {
  @PrimaryGeneratedColumn()
  id: number;

  /** `name` is for equivalent enrichment's reference db name */
  @Index({ unique: true })
  @Column({ type: 'text' })
  name: string;
}

@Entity({ schema: 'public' })
export class EnrichmentTerm {
  @PrimaryGeneratedColumn()
  id: number;

  /** `id` is for equivalent enrichment's reference source id */
  @Index({ unique: true })
  @Column({ type: 'text' })
  enrichmentId: string;

  /** `name` is for equivalent enrichment's reference source name */
  @Column({ type: 'text' })
  enrichmentName: string;

  @Column({ type: 'int' })
  enrichmentDbId: number;

  @ManyToOne(() => EnrichmentDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'enrichment_db_id' })
  enrichmentDB: EnrichmentDB;
}

@Entity({ schema: 'public' })
@Index(['molecularDbId', 'formula'])
export class EnrichmentDBMoleculeMapping {
  @PrimaryGeneratedColumn()
  id: number;

  /** molecular db `name` for molecule to map with reference enrichment db */
  @Column({ type: 'text' })
  moleculeEnrichedName: string;

  @Column({ type: 'text' })
  formula: string;

  @Column({ type: 'int' })
  enrichmentTermId: number;

  @Column({ type: 'int' })
  moleculeId: number;

  @Column({ type: 'int' })
  molecularDbId: number;

  @ManyToOne(() => EnrichmentTerm, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'enrichment_term_id' })
  enrichmentTerm: EnrichmentTerm;

  @ManyToOne(() => MolecularDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'molecular_db_id' })
  molecularDB: MolecularDB;
}

@Entity({ schema: 'public' })
export class EnrichmentBootstrap {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: 'int' })
  scenario: number;

  @Column({ type: 'text' })
  formulaAdduct: string;

  @Column({ name: 'dataset_id' })
  datasetId: string;

  @Column({ name: 'annotation_id' })
  annotationId: string;

  @Column({ type: 'numeric', precision: 2, scale: 2 })
  fdr: string;

  @Column({ type: 'int' })
  enrichmentDbMoleculeMappingId: number;

  // This table is a child table of public.dataset, not graphql.dataset, so avoid making an FK
  @ManyToOne(() => Dataset, { createForeignKeyConstraints: false })
  @JoinColumn({ name: 'dataset_id' })
  dataset: Dataset;

  @ManyToOne(() => EngineDataset, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'dataset_id' })
  engineDataset: EngineDataset;

  @ManyToOne(() => EnrichmentDBMoleculeMapping, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'enrichment_db_molecule_mapping_id' })
  enrichmentDBMoleculeMapping: EnrichmentDBMoleculeMapping;
}

@Entity({ schema: 'public' })
export class DatasetEnrichment {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'dataset_id' })
  datasetId: string;

  @Column({ type: 'int' })
  enrichmentDbId: number;

  @Column({ type: 'timestamp without time zone', nullable: true, transformer: new MomentValueTransformer() })
  processingDt: Date | null;

  @Column({ type: 'int' })
  molecularDbId: number;

  // This table is a child table of public.dataset, not graphql.dataset, so avoid making an FK
  @ManyToOne(() => Dataset, { createForeignKeyConstraints: false })
  @JoinColumn({ name: 'dataset_id' })
  dataset: Dataset;

  @ManyToOne(() => MolecularDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'molecular_db_id' })
  molecularDB: MolecularDB;

  @ManyToOne(() => EnrichmentDB, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'enrichment_db_id' })
  enrichmentDB: EnrichmentDB;
}

export const ENRICHMENT_DB_ENTITIES = [
  EnrichmentDB,
  EnrichmentTerm,
  EnrichmentDBMoleculeMapping,
  EnrichmentBootstrap,
  DatasetEnrichment,
]
