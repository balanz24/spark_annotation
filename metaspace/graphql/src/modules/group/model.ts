import {
  Entity,
  PrimaryColumn,
  Column,
  JoinColumn,
  OneToMany,
  ManyToOne,
} from 'typeorm'

import { User } from '../user/model'
import { UserGroupRole } from '../../binding'
import { MolecularDB } from '../moldb/model'

export const UserGroupRoleOptions: Record<UserGroupRole, UserGroupRole> = {
  INVITED: 'INVITED',
  PENDING: 'PENDING',
  MEMBER: 'MEMBER',
  GROUP_ADMIN: 'GROUP_ADMIN',
}

@Entity()
export class Group {
  @PrimaryColumn({ type: 'uuid', default: () => 'uuid_generate_v1mc()' })
  id: string;

  @Column({ type: 'text' })
  name: string;

  @Column({ type: 'text' })
  shortName: string;

  @Column({ type: 'text', nullable: true })
  urlSlug: string | null;

  @Column({ type: 'text', name: 'group_description', default: '' })
  groupDescriptionAsHtml: string;

  @OneToMany(() => UserGroup, userGroup => userGroup.group)
  members: UserGroup[];

  @OneToMany(() => GroupDetectability, groupDetectability => groupDetectability.group)
  sources: GroupDetectability[];

  @OneToMany(() => MolecularDB, molecularDB => molecularDB.group)
  molecularDBs: MolecularDB[];
}

@Entity('user_group')
export class UserGroup {
  @PrimaryColumn({ type: 'uuid' })
  userId: string;

  @ManyToOne(() => User)
  @JoinColumn({ name: 'user_id' })
  user: User;

  @PrimaryColumn({ type: 'uuid' })
  groupId: string;

  @ManyToOne(() => Group)
  @JoinColumn({ name: 'group_id' })
  group: Group;

  @Column({ type: 'text', enum: ['INVITED'] })
  role: 'INVITED' |
    'PENDING' |
    'MEMBER' |
    'GROUP_ADMIN';

  @Column({ default: true })
  primary: boolean;
}
@Entity('group_detectability')
export class GroupDetectability {
  @PrimaryColumn({ type: 'uuid' })
  id: string;

  @PrimaryColumn({ type: 'uuid' })
  groupId: string;

  @ManyToOne(() => Group)
  @JoinColumn({ name: 'group_id' })
  group: Group;

  @Column({ type: 'text', enum: ['EMBL'] })
  source: 'EMBL' |
    'ALL' |
    'INTERLAB';
}

export const GROUP_ENTITIES = [
  Group,
  UserGroup,
  GroupDetectability,
]
