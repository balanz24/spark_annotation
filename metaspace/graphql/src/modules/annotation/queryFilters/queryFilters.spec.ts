import * as _ from 'lodash'
import {
  anonContext,
  onAfterAll,
  onAfterEach,
  onBeforeAll,
  onBeforeEach,
  testEntityManager,
} from '../../../tests/graphqlTestEnvironment'
import { ColocAnnotation, ColocJob, Ion } from '../model'
import { applyQueryFilters, ESAnnotationWithColoc } from './index'
import { ESAnnotation } from '../../../../esConnector'
import { DeepPartial } from 'typeorm'

import { QueryFilterArgs } from './types'
import * as moment from 'moment'
import { MolecularDB } from '../../moldb/model'
import { createTestMolecularDB } from '../../../tests/testDataCreation'

describe('annotation/queryFilters applyQueryFilters (colocalization)', () => {
  beforeAll(onBeforeAll)
  afterAll(onAfterAll)
  // beforeEach(onBeforeEach);
  afterEach(onAfterEach)

  let ions: Ion[]
  let job: ColocJob
  let database: MolecularDB

  beforeEach(async() => {
    await onBeforeEach()

    ions = testEntityManager.create(Ion, [1, 2, 3, 4].map(i => ({
      ion: `H${i}O+H+`,
      formula: `H${i}O`,
      chemMod: i & 1 ? '-H+OH' : '',
      neutralLoss: i & 2 ? '-OH' : '',
      adduct: '+H',
      charge: 1,
    })))
    await testEntityManager.insert(Ion, ions)

    database = await createTestMolecularDB({ name: 'HMDB-v4', version: 'version' })

    job = testEntityManager.create(ColocJob, {
      datasetId: 'datasetId',
      moldbId: database.id,
      fdr: 0.1,
      algorithm: 'cosine',
      sampleIonIds: [ions[1].id, ions[2].id],
      start: moment(),
      finish: moment(),
    })
    await testEntityManager.insert(ColocJob, job)

    const annotations = testEntityManager.create(ColocAnnotation, _.range(ions.length).map(i => ({
      colocJobId: job.id,
      ionId: ions[i].id,
      colocIonIds: [...ions, ...ions].slice(i + 1, i + 3).map(ion => ion.id),
      colocCoeffs: [0.9, 0.7],
    })))
    await testEntityManager.insert(ColocAnnotation, annotations)
  })

  it('should transform the args of a request with a colocalizedWith filter', async() => {
    const argsWithColocWith: QueryFilterArgs = {
      datasetFilter: { ids: job.datasetId },
      filter: {
        databaseId: job.moldbId,
        fdrLevel: job.fdr,
        colocalizedWith: ions[1].ion,
        colocalizationAlgo: job.algorithm,
      },
    }

    const { args } = await applyQueryFilters(anonContext, argsWithColocWith)

    expect(args).toMatchObject({
      offset: 0,
      limit: 1000,
      filter: {
        databaseId: database.id,
        ion: ions.slice(1, 4).map(ion => ion.ion),
      } as any,
    })
    args.filter!.databaseId = 1
    expect(args).toMatchSnapshot()
  })

  it('should transform the args of a request with a colocalizationSamples filter', async() => {
    const argsWithColocWith: QueryFilterArgs = {
      datasetFilter: { ids: job.datasetId },
      filter: {
        databaseId: job.moldbId,
        fdrLevel: job.fdr,
        colocalizationSamples: true,
        colocalizationAlgo: job.algorithm,
      },
    }

    const { args } = await applyQueryFilters(anonContext, argsWithColocWith)

    expect(args).toMatchObject({
      filter: {
        databaseId: database.id,
        ion: ions.slice(1, 3).map(ion => ion.ion),
      } as any,
    })
    args.filter!.databaseId = 1
    expect(args).toMatchSnapshot()
  })

  it('should add colocalizationCoeffs to annotations', async() => {
    const argsWithColocWith: QueryFilterArgs = {
      datasetFilter: { ids: job.datasetId },
      filter: {
        databaseId: job.moldbId,
        fdrLevel: job.fdr,
        colocalizedWith: ions[1].ion,
        colocalizationAlgo: job.algorithm,
      },
    }
    const ionAnnotations = ions.map(({ ion, formula, chemMod, neutralLoss, adduct }) =>
      ({ ion, formula, chem_mod: chemMod, neutral_loss: neutralLoss, adduct }))
    const { postprocess } = await applyQueryFilters(anonContext, argsWithColocWith)
    const annotations: DeepPartial<ESAnnotation>[] = [
      { _source: ionAnnotations[2] },
      { _source: ionAnnotations[3] },
      { _source: ionAnnotations[1] },
    ]

    const result = postprocess!(annotations as ESAnnotation[]) as ESAnnotationWithColoc[]

    expect(result).toMatchObject([
      { _source: ionAnnotations[1], _isColocReference: true },
      { _source: ionAnnotations[2], _isColocReference: false },
      { _source: ionAnnotations[3], _isColocReference: false },
    ])
    expect(result[0].getColocalizationCoeff(ions[1].ion, job.algorithm, job.moldbId, job.fdr)).toEqual(1)
    expect(result[1].getColocalizationCoeff(ions[1].ion, job.algorithm, job.moldbId, job.fdr)).toEqual(0.9)
    expect(result[2].getColocalizationCoeff(ions[1].ion, job.algorithm, job.moldbId, job.fdr)).toEqual(0.7)
  })
})
