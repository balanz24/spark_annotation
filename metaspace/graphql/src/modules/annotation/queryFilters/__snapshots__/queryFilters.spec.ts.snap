// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`annotation/queryFilters applyQueryFilters (colocalization) should transform the args of a request with a colocalizationSamples filter 1`] = `
Object {
  "datasetFilter": Object {
    "ids": "datasetId",
  },
  "filter": Object {
    "colocalizationAlgo": "cosine",
    "colocalizationSamples": true,
    "databaseId": 1,
    "fdrLevel": 0.1,
    "ion": Array [
      "H2O+H+",
      "H3O+H+",
    ],
  },
}
`;

exports[`annotation/queryFilters applyQueryFilters (colocalization) should transform the args of a request with a colocalizedWith filter 1`] = `
Object {
  "datasetFilter": Object {
    "ids": "datasetId",
  },
  "filter": Object {
    "colocalizationAlgo": "cosine",
    "colocalizedWith": "H2O+H+",
    "databaseId": 1,
    "fdrLevel": 0.1,
    "ion": Array [
      "H2O+H+",
      "H3O+H+",
      "H4O+H+",
    ],
  },
  "limit": 1000,
  "offset": 0,
}
`;
